### DISCUSSION TITLE
DB connect kills the process on error

### DISCUSSION THREAD

**Discussion by c-wygoda** [payloadcms/payload#3784] [Feature Requests & Ideas]
### Describe the Bug

It is not possible to use the `db.connect` method with retrying on error, as it catches the connection errors and kills the process:
- https://github.com/payloadcms/payload/blob/main/packages/db-postgres/src/connect.ts#L31
- https://github.com/payloadcms/payload/blob/main/packages/db-mongodb/src/connect.ts#L59

IMHO, instead of killing there, the error should be allowed to bubble up and caught and possibly process-killed in Payload's `init` default connection attempt: https://github.com/payloadcms/payload/blob/main/packages/payload/src/payload.ts#L360

This would allow to set `disableDBConnect` to `false` and for example try and retry the DB connection in the `onInit` handler.

My motivation is to start a dockerized DB in parallel to Payload and just retry until it is available. Using something like wait-for-it.sh always feels like code not handling unready resources very well to me. I might even argue for the default connect attempt in the init method to retry, maybe with an max timeout to allow waiting for starting databases.

### To Reproduce

- Start Payload without a running DB, even with retrying `payload.db.connect()` on error in the onInit handler with `disableDBConnect` set to false.

### Payload Version

2

### Adapters and Plugins

db-postgres

**Comment by jmikrut** [comment]
Hey @c-wygoda I think this is a reasonable request. We could attempt to handle connections as you've noted, and then only throw after a certain amount of failed attempts. 

Have you done this successfully with other packages in the past? Say, a raw ORM or similar? 

I will add this as a roadmap candidate. For now, you could ensure database availability before you call `payload.init()`, but I could see how that would be cumbersome and it would be more graceful to have Payload handle it for you.

I'll keep you updated regarding the status of this feature request on our roadmap!

**Comment by cwygoda** [comment]
Here's a draft PR with what I think this should like (sorry for account jumping): https://github.com/payloadcms/payload/pull/3790

I'm using retries across my projects - or rather have my teams use it - when talking to external resources as a defensive measure against whatever might be flaky - DNS, networking in general, still starting, etc. Usually with an exponential backoff, but I think this is a decent start.

Would you point me to where to add tests for this and documentation? I hope to find some time in the next days after vacation to provide those changes as well.
